File: leetcode/datastructure/stack/stack.java
package leetcode.datastructure.stack;
import java.util.EmptyStackException;
class stack {
    int MAX;
    int[] stackArray;
    int top;
    public stack(int size) {
        this.MAX = size;
        this.stackArray = new int[MAX];
        this.top = -1;
    }

    private boolean isEmpty() {
        return top < 0;
    }

    private boolean isFull() {
        return top == MAX - 1;
    }

    private void resize() {
        int newSize = MAX * 2;
        int[] newArray = new int[newSize];
        for (int i = 0; i < MAX; i++) {
            newArray[i] = stackArray[i];
        }
        this.stackArray = newArray;
        this.MAX = newSize;
    }

    private void push(int val) {
        if (!isFull()) {
            this.stackArray[++this.top] = val;
        }
        else {
            resize();
            this.stackArray[++this.top] = val;
        }
    }

    private int pop() {
        if (!isEmpty()) {
            return this.stackArray[this.top--];
        }
        else {
            throw new EmptyStackException();
        }
    }

    private int peek() {
        if (!isEmpty()) {
            return this.stackArray[this.top];
        }
        else {
            throw new EmptyStackException();
        }
    }

    private void display() {
        String str = "";
        int cnt = this.top;
        while (cnt != -1) {
            str += this.stackArray[cnt--];
            str += "->";
        }
        str += "null";
        System.out.println(str);
    }

    public static void main(String[] args) {
        stack stackArray = new stack(10);
        for (int i = 0; i < 12; i++) {
            stackArray.push(i);
        }
        stackArray.display();
        System.out.println(stackArray.pop() + "->" + stackArray.peek());
    }
}

File: leetcode/datastructure/stack/stack.py
# coding=utf-8

class stack(object):
    def __init__(self, val):
        self.MAX = val
        self.stackArray = []
        self.top = -1

    def isEmpty(self):
        return self.top < 0

    def isFull(self):
        return self.top == self.MAX - 1

    def push(self, val):
        if self.isFull():
            self.MAX = self.MAX * 2
        self.top = self.top + 1
        self.stackArray.append(val)

    def pop(self):
        if not self.isEmpty():
            self.top = self.top - 1
            return self.stackArray.pop()
        raise Exception('empty')
            
    def peek(self):
        if not self.isEmpty():
            return self.stackArray[self.top]
        raise Exception('empty')

    def display(self):
        pstr = ''
        for i in range(self.top, -1, -1):
            pstr = pstr + str(self.stackArray[i])
            pstr = pstr + '->'
        pstr = pstr + 'None'
        print pstr

if __name__ == "__main__":
    s = stack(10)
    for i in range(20):
        s.push(i)
    s.display()
    print s.peek()
    print s.pop()
    print s.peek()